crates/subspace-farmer/src/single_plot_farm.rs

```diff
@@ -25,10 +25,51 @@ use subspace_solving::{BatchEncodeError, SubspaceCodec};
 use thiserror::Error;
 use tokio::runtime::Handle;
 use tokio::task::JoinHandle;
-use tracing::{error, info};
+use tracing::{error, info, trace};
 
 const SYNC_PIECES_AT_ONCE: u64 = 5000;
 
+#[derive(Debug, Clone)]
+pub struct SinglePlotPieceGetter {
+    codec: SubspaceCodec,
+    plot: Plot,
+}
+
+impl SinglePlotPieceGetter {
+    pub fn new(codec: SubspaceCodec, plot: Plot) -> Self {
+        Self { codec, plot }
+    }
+
+    pub fn get_piece(
+        &self,
+        piece_index: PieceIndex,
+        piece_index_hash: PieceIndexHash,
+    ) -> Option<Piece> {
+        match self.plot.read_piece(piece_index_hash) {
+            Ok(mut piece) => match self.codec.decode(&mut piece, piece_index) {
+                Ok(()) => {
+                    return Some(piece);
+                }
+                Err(error) => {
+                    trace!(
+                        %error,
+                        "Piece with piece index {piece_index} not found in plot"
```

Comment:
This is definitely not right. We might fail here only if the decoding fails, which might happen due to gpu issues.

Reply:
Yes, fixed